src/nbl/asset/interchange/CIESProfileLoader.cpp

```diff
@@ -0,0 +1,352 @@
+#include "nbl/asset/interchange/CIESProfileLoader.h"
+
+using namespace nbl;
+using namespace asset;
+
+
+namespace nbl {
+namespace asset {
+
+class CIESProfile {
+public:
+    _NBL_STATIC_INLINE_CONSTEXPR double MAX_VANGLE = 180.0;
+    _NBL_STATIC_INLINE_CONSTEXPR double MAX_HANGLE = 360.0;
+
+    enum PhotometricType : uint32_t {
+        TYPE_NONE,
+        TYPE_C,
+        TYPE_B,
+        TYPE_A,
+    };
+
+    CIESProfile() = default;
+    CIESProfile(PhotometricType type, size_t hSize, size_t vSize)
+        : type(type), hAngles(hSize), vAngles(vSize), data(hSize* vSize) {}
+    ~CIESProfile() = default;
+    core::vector<double>& getHoriAngles() { return hAngles; }
+    const core::vector<double>& getHoriAngles() const { return hAngles; }
+    core::vector<double>& getVertAngles() { return vAngles; }
+    const core::vector<double>& getVertAngles() const { return vAngles; }
+    void addHoriAngle(double hAngle) {
+        hAngles.push_back(hAngle);
+        data.resize(getHoriSize() * getVertSize());
+    }
+    size_t getHoriSize() const { return hAngles.size(); }
+    size_t getVertSize() const { return vAngles.size(); }
+    void setValue(size_t i, size_t j, double val) {
+        data[getVertSize() * i + j] = val;
+    }
+    double getValue(size_t i, size_t j) const {
+        return data[getVertSize() * i + j];
+    }
+    double sample(double vAngle, double hAngle) const;
+    double getMaxValue() const {
+        return *std::max_element(std::begin(data), std::end(data));
+    }
+    double getIntegral() const;
+private:
+    PhotometricType type;
+    core::vector<double> hAngles;
+    core::vector<double> vAngles;
+    core::vector<double> data;
+};
+
+}
+}
+
+class CIESProfileParser {
+public:
+    CIESProfileParser(char* buf, size_t size) { ss << std::string(buf, size); }
+
+    bool parse(CIESProfile& result);
+
+private:
+    int getInt(const char* errorMsg) {
+        int in;
+        if (ss >> in)
+            return in;
+        error = true;
+        if (!this->errorMsg)
+            this->errorMsg = errorMsg;
+        return 0;
+    }
+
+    double getDouble(const char* errorMsg) {
+        double in;
+        if (ss >> in)
+            return in;
+        error = true;
+        if (!this->errorMsg)
+            this->errorMsg = errorMsg;
+        return -1.0;
+    }
+
+    bool error{ false };
+    const char* errorMsg{ nullptr };
+    std::stringstream ss;
+};
+
+double CIESProfile::sample(double vAngle, double hAngle) const {
+    vAngle = fmod(vAngle, vAngles.back()); // warp angle
+    hAngle = hAngles.back() == 0.0
+                 ? 0.0
+                 : fmod(hAngle,
+                        hAngles.back()); // when last horizontal angle is zero
```

Comment:
`fmod` gives you weird warping for vertical angles, I'd rather clamp when the range of the values isn't 0,2PI

Reply:
vAngles.back() can be 90 degrees or 180 degrees. Warping is explicitly needed here. I can clamp vAngle to 0,2pi though.

Reply:
> vAngles.back() can be 90 degrees or 180 degrees. Warping is explicitly needed here. I can clamp vAngle to 0,2pi though.

If you're over 180 you're wrapping around over the south pole of the sphere, and then you end up flipping your horizontal angle.

Also for 90 degs, fmod will make you wrap around to a higher angle so the value at 0 will repeat just past, 90.

You probably want MIRROR wrapping here, not REPEAT

Reply:
fixed in #660 it was actually CLAMP_TO_BORDER with border being black thats correct